### ISSUE TITLE
Is it possible to test the data that isn't in the train-set? New images with pre-trained model.

### ISSUE THREAD

**Issue by wjun0830** [yunjey/stargan#108]
Is it possible to test the data with pre-trained model?

I am curious if there is any way to test new images with the model I trained in the past such as training with only celebA but later testing with custom datasets.

**Comment by ujjawalcse** [comment]
I too wanna know how to test on our custom datasets with the pre-trained model.
Have you tested @wjun0830?

**Comment by wjun0830** [comment]
Nope not yet. I think i have to modify the code if i want to test the whole
custom set since testset is already specified in Class when training.
However, by swapping a image in celeba result direc with custom image, I
could test it on a single image.

2019년 11월 15일 (금) 오후 6:27, Ujjawal Kumar Singh <notifications@github.com>님이
작성:

**Comment by ujjawalcse** [comment]
Thank you Wjun.
But I tested it as I used to test with normal gan. As we have the
checkpoints of the generator, we can load the checkpoints after defining
the generator model. Then use it for new image generation. By the way, I
couldn't get the final generated image due to my Runtime Error: CUDA out of
memory. But you can try it. Most probably, it will give the result.  This
is the code:-

import torch
import cv2
import numpy as np
import model
import solver
from torch.backends import cudnn
from torchvision.utils import save_image
from PIL import Image

model = model.Generator()
model = model.cuda()
model.load_state_dict(torch.load("./stargan_celeba_256/models/200000-G.ckpt"))
#model = model.eval()

pilImg = Image.open("data/celeba/test/input_1573800763_075695.jpg")
npImg = np.array(pilImg)
#npImg=cv2.imread('data/celeba/test/input_1573800763_075695.jpg')
#npImg=cv2.cvtColor(npImg,cv2.COLOR_BGR2RGB)
#npImg=cv2.resize(npImg,(256,256),interpolation=cv2.INTER_AREA)
example_img = torch.from_numpy(npImg).float().unsqueeze(0)
print(example_img.shape)
example_img = example_img.permute(0, 3, 1, 2)
print(example_img.shape)
print("======RGB data [0, 255]======")
print(example_img)
example_img = example_img.div(255)
example_img = example_img.sub(0.5)
example_img = example_img.div(0.5)
example_img = example_img.cuda()

example_c = torch.cuda.FloatTensor([[0,0,1,0,0]])

imgGen = model(example_img, example_c)
save_image(imgGen, "./%s.png" % 100, normalize=True)

**Comment by ujjawalcse** [comment]
I checked with this code on another machine, but the result is not good.

**Comment by wjun0830** [comment]
Oh thanks.
Can I see the results? I thought stargan needs the picture with attributes
array so this code may not work. when you see the torch code in its github,
they preprocess image with the matching attributes.
Did you consider that?



2019년 11월 15일 (금) 오후 11:59, Ujjawal Kumar Singh <notifications@github.com>님이
작성:

**Comment by wjun0830** [comment]
Ujjawal, Could you tell me which package should i install to get the
package 'model'?

I am not the supervisor in my server and I should request to install some
models but when I try to install 'model' it fails.

import model
doesnt work for me, so I tried pip install model but there is no matching
distribution. Can u tell me what the package name is?

2019년 11월 16일 (토) 오전 2:36, 문원준 <wjun0830@gmail.com>님이 작성:

**Comment by ujjawalcse** [comment]
You have whole folder of stargan-pytorch. In this folder there are 8-9
files.
There are model.py,solver.py, data_loader.py etc. So you can these files
as:-
import solver
import model
import data_loader.py
etc.
model is not inbuilt package so it can't be installed by pip install model.
It's just a python file.

**Comment by wjun0830** [comment]
Oh, I missed that. and also the tensor in the code. Sorry..
Sorry but I have no idea how to improve that.
Maybe I will keep this issue open and wait..

**Comment by wjun0830** [comment]
Maybe You will wanna try this way. This result is also not good but maybe a hint for you.
I changed the function test in original code in solver.py.

here is the code
I utilized your code into the original code and I still do not know why this is not working. 
'''

    def test(self):
        """Translate images using StarGAN trained on a single dataset."""
        # Load the trained generator.
        self.restore_model(self.test_iters)

        # Set data loader.
        if self.dataset == 'CelebA':
            data_loader = self.celeba_loader
        elif self.dataset == 'RaFD':
            data_loader = self.rafd_loader

        with torch.no_grad():
            i = 0;
            pilImg = Image.open("data/celeba/images/028136.jpg")
            npImg = np.array(pilImg)
            example_img = torch.from_numpy(npImg).float().unsqueeze(0)
            example_img = example_img.permute(0,3,1,2)
            example_img = example_img.div(255)
            example_img = example_img.sub(0.5)
            example_img = example_img.div(0.5)
            x_real = example_img.cuda()
            c_org = torch.cuda.FloatTensor([[1,-1,-1,1,1,-1,-1,-1]])
            x_real = x_real.to(self.device)
            x_fake_list = [x_real]
            c_trg_list = self.create_labels(c_org, self.c_dim, self.dataset, self.selected_attrs)
            for c_trg in c_trg_list:
              x_fake_list.append(self.G(x_real, c_trg))
            for num, k in enumerate(x_fake_list):
              result_path = os.path.join(self.result_dir, '{}-images.jpg'.format((i)*9+num))
              save_image(self.denorm(k.data.cpu()), result_path, nrow=1, padding=0)
              print('Saved real and fake images into {}...'.format(result_path))

'''

**Comment by ujjawalcse** [comment]
c_org should be only a list of 5 items because the model is trained on 5
attributes only.
You can put value 0 or 1 based on which attribute you wanna get as a
result.
For example [0,0,1,0,0] for brown hair and [0,1,0,0,0] for Blond hair.

**Comment by wjun0830** [comment]
Oh sorry didnt change that for you
I trained with 8attributes and thats why the tensor shape is like that

2019년 11월 18일 (월) 오후 2:59, Ujjawal Kumar Singh <notifications@github.com>님이
작성:

> c_org should be only a list of 5 items because the model is trained on 5
> attributes only.
> You can put value 0 or 1 based on which attribute you wanna get as a
> result.
> For example [0,0,1,0,0] for brown hair and [0,1,0,0,0] for Blond hair.
>
>

**Comment by wjun0830** [comment]
And also if you look at celeba attr file, it has 1 and -1 values and that
is why I changed the value

2019년 11월 18일 (월) 오후 3:19, 문원준 <wjun0830@gmail.com>님이 작성:

**Comment by wjun0830** [comment]
I have solved the problem.

https://github.com/wjun0830/stargan
I have uploaded the code here.
You will see the code in line 525 in solver2.py and will also need to include the header transforms.

**Comment by mumuyanyan** [comment]
> I have solved the problem.
> 
> https://github.com/wjun0830/stargan
> I have uploaded the code here.
> You will see the code line 525 in solver2.py and will also need to include the header transforms.

thanks for your contribution. but my results are very bad.
what are the results of you?
![0-images](https://user-images.githubusercontent.com/13342178/71355442-010d1200-25ba-11ea-9216-ebe8c3b70978.jpg)
![1-images](https://user-images.githubusercontent.com/13342178/71355447-023e3f00-25ba-11ea-8fbf-ff62778f37c1.jpg)
![2-images](https://user-images.githubusercontent.com/13342178/71355449-036f6c00-25ba-11ea-967f-be595d958140.jpg)
![3-images](https://user-images.githubusercontent.com/13342178/71355450-04a09900-25ba-11ea-9b40-677f7d08b562.jpg)
![4-images](https://user-images.githubusercontent.com/13342178/71355452-066a5c80-25ba-11ea-809b-2e597427f0ad.jpg)
![5-images](https://user-images.githubusercontent.com/13342178/71355454-079b8980-25ba-11ea-9b0a-063bcf600ac0.jpg)

**Comment by wjun0830** [comment]
![image](https://user-images.githubusercontent.com/31557552/71358870-d4fb8c00-25cd-11ea-9eb1-eb0111633f1d.png)

The result may depends on the pictures. When I tried another picture of old and bald man, the result was similar to your output.
